### PR TITLE
support for additional types

### PR DESCRIPTION
**Pull request by bionoren**

### CODE REVIEW

File: encoder.go [hunk 1]
```diff
@@ -58,8 +58,28 @@ func (e *encoder) encode(v any) {
...
+	case int8:
+		e.w.Write(strconv.AppendInt(e.buf[:0], int64(v), 10))
+	case int16:
+		e.w.Write(strconv.AppendInt(e.buf[:0], int64(v), 10))
+	case int32:
+		e.w.Write(strconv.AppendInt(e.buf[:0], int64(v), 10))
+	case int64:
+		e.w.Write(strconv.AppendInt(e.buf[:0], v, 10))
+	case float32:
+		e.encodeFloat64(float64(v))
```
Comment (nathanmcgarvey-modopayments) [inline encoder.go:82]: Won't this have rounding errors?

File: func.go [hunk 2]
```diff
@@ -977,7 +977,7 @@ func funcIndex2(_, v, x any) any {
 				return &expectedObjectError{v}
 			}
 		}
-	case int, float64, *big.Int:
+	case int, int8, int16, int32, int64, float32, float64, *big.Int:
```
Comment (nathanmcgarvey-modopayments) [inline func.go]: This isn't all this file needs. I was tinkering earlier and you might need this in the `toInt()`:

```	
case int8:
		return int(x), true // We assume that all platforms are at least 32 bit
	case int16:
		return int(x), true // We assume that all platforms are at least 32 bit
	case int32:
		return int(x), true // We assume that all platforms are at least 32 bit
	case int64:
		if uintSize >= 64 || x <= 2147483647 {
			return int(x), true
		}
		return int(x), false // Overflow

	case uint:
		if x <= 2147483647 || (uintSize == 64 && x <= 9223372036854775807) {
			return int(x), true
		}
		return math.MaxInt, false // Overflow
	case uint8:
		return int(x), true // We assume that all platforms are at least 32 bit
	case uint16:
		return int(x), true // We assume that all platforms are at least 32 bit
	case uint32:
		if uintSize > 32 || x <= 2147483647 {
			return int(x), true
		}
		return math.MaxInt, false // Overflow
	case uint64:
		if uintSize == 64 && x <= 9223372036854775807 {
			return int(x), true
		}
		return math.MaxInt, false // Overflow```

Comment (nathanmcgarvey-modopayments) [reply func.go]: Maybe similar in `toFloat` (for math operations)

**Review by nathanmcgarvey-modopayments — CHANGES_REQUESTED** [review]

File: func.go (hunk 2 above)
Comment (nathanmcgarvey-modopayments) [reply func.go]: Test case would be to use the `funcHas` or `slice` or `sliceString` functions as they take the `toInt` code path
toFloat is called by things like `funcFlatten` or `Isnan` or some others.

Comment (nathanmcgarvey-modopayments) [reply func.go]: After looking at it, that func.go file is a mess with these:

Take a look at `funcAbs`, or `funcLength` for example)

*sigh*. Thoughts?

Comment (nathanmcgarvey-modopayments) [reply func.go]: Even worse: `operator.go`... checkout `funcOpAdd` or most of the stuff in there, for example.

File: encoder.go (hunk 1 above)
Comment (bionoren) [reply encoder.go:82]: we're up-converting (32->64), so it should be safe

Comment (nathanmcgarvey-modopayments) [reply encoder.go:82]: Odd:

https://go.dev/play/p/YgXowJ149Bh

Looks like a full round trip is safe, but one way may not be. Not sure how much we care, or if "rounding error" is just expected at this point.

**Review by nathanmcgarvey-modopayments — APPROVED** [review]

**Pull request merged**